.github/workflows/common.main.kts

```diff
@@ -0,0 +1,159 @@
+import io.github.typesafegithub.workflows.dsl.expressions.Contexts.secrets
+import io.github.typesafegithub.workflows.dsl.expressions.expr
+import java.util.Properties
+
+val SPOCK_BUILD_CACHE_USERNAME by secrets
+val SPOCK_BUILD_CACHE_PASSWORD by secrets
+val GRADLE_ENTERPRISE_ACCESS_KEY by secrets
+
+val commonCredentials = linkedMapOf(
+    "ORG_GRADLE_PROJECT_spockBuildCacheUsername" to expr(SPOCK_BUILD_CACHE_USERNAME),
+    "ORG_GRADLE_PROJECT_spockBuildCachePassword" to expr(SPOCK_BUILD_CACHE_PASSWORD),
+    "GRADLE_ENTERPRISE_ACCESS_KEY" to expr(GRADLE_ENTERPRISE_ACCESS_KEY)
+)
+
+data class Strategy(
+    val failFast: Boolean? = false,
+    val matrix: Matrix? = null
+) {
+    fun toCustomArguments() = mapOf(
+        *listOfNotNull(
+            failFast?.let { "fail-fast" to failFast },
+            matrix?.let { "matrix" to matrix.toCustomArguments() }
+        ).toTypedArray()
+    )
+}
+
+data class Matrix(
+    val operatingSystems: List<String>? = null,
+    val variants: List<String>? = null,
+    val javaVersions: List<String>? = null,
+    val excludes: List<Map<String, String>>? = null,
+    val includes: List<Map<String, String>>? = null
+) {
+    fun toCustomArguments() = mapOf(
+        *listOfNotNull(
+            operatingSystems?.let { "os" to operatingSystems },
+            variants?.let { "variant" to variants },
+            javaVersions?.let { "java" to javaVersions },
+            excludes?.let { "exclude" to excludes },
+            includes?.let { "include" to includes }
+        ).toTypedArray()
+    )
+
+    data class Axes(
+        val javaVersions: List<String>,
+        val variants: List<String>
+    )
+
+    companion object {
+        val operatingSystem = "matrix.os"
+        val variant = "matrix.variant"
+        val java = "matrix.java"
+    }
+}
+
+fun WorkflowBuilder.job(
+    id: String,
+    name: String? = null,
+    runsOn: RunnerType,
+    needs: List<Job<*>> = emptyList(),
+    condition: String? = null,
+    strategy: Strategy? = null,
+    simpleStrategy: Map<String, List<String>>? = null,
+    block: JobBuilder<EMPTY>.() -> Unit
+): Job<EMPTY> = job(
+    id = id,
+    name = name,
+    runsOn = runsOn,
+    needs = needs,
+    condition = condition,
+    strategyMatrix = simpleStrategy,
+    _customArguments = mapOf(
+        *listOfNotNull(
+            strategy?.let { "strategy" to strategy.toCustomArguments() }
+        ).toTypedArray()
+    ),
+    block = block
+)
+
+val Matrix.Companion.full
+    get() = Matrix(
+        operatingSystems = listOf("ubuntu-latest"),
+        variants = axes.variants,
+        javaVersions = axes.javaVersions,
+        excludes = axes.javaVersions
+            .filter { it.toInt() >= 17 }
+            .map { javaVersion ->
+                mapOf(
+                    "os" to "ubuntu-latest",
+                    "variant" to "2.5",
+                    "java" to javaVersion
+                )
```

Comment:
Why do we not use a `data class` for these?

Reply:
Sure, why not. :-)